### ISSUE TITLE
Use native MacOS DNS resolution

### ISSUE THREAD

**Issue by dmikusa-pivotal** [cloudfoundry/cli#1433]
## Command

N/A

## What occurred

DNS resolution does not use the native MacOS resolution. Thus it does not take into account `/etc/resolver/`.

## What you expected to occur

Use native MacOS DNS resolution by default or make it adjustable via env variable.  [This does not seem to work with the present binaries either](https://godoc.org/net#hdr-Name_Resolution).

## CLI Version

6.37

## CC API Endpoint Version

N/A

## CF Trace

Defaults to Go DNS resolver.

```
$ GODEBUG=netdns=2 ./cf apps
Getting apps in org labs / space dmikusa as admin...
go package net: built with netgo build tag; using Go's DNS resolver
go package net: hostLookupOrder(api.run-04.haas-59.pez.pivotal.io) = files,dns
go package net: hostLookupOrder(api.run-04.haas-59.pez.pivotal.io) = files,dns
go package net: hostLookupOrder(api.run-04.haas-59.pez.pivotal.io) = files,dns
FAILED
```

Cannot override to use `cgo`.

```
$ GODEBUG=netdns=cgo+2 ./cf apps
Getting apps in org labs / space dmikusa as admin...
go package net: built with netgo build tag; using Go's DNS resolver
go package net: hostLookupOrder(api.run-04.haas-59.pez.pivotal.io) = files,dns
go package net: hostLookupOrder(api.run-04.haas-59.pez.pivotal.io) = files,dns
go package net: hostLookupOrder(api.run-04.haas-59.pez.pivotal.io) = files,dns
FAILED
```

## Platform & Shell Details

Mac OS 10.13.6, iTerm2 3.2.0

## Any other relevant information

If I compile my own binary and set `REQUIRED_FOR_STATIC_BINARY =-a -tags cgo -installsuffix cgo` and `CGO_ENABLED=1 GOARCH=amd64 GOOS=darwin go build` in the Makefile, it works using the MacOS DNS resolution.

```
$ GODEBUG=netdns=2 cf apps
Getting apps in org labs / space dmikusa as admin...
go package net: using cgo DNS resolver
go package net: hostLookupOrder(api.run-04.haas-59.pez.pivotal.io) = cgo
go package net: hostLookupOrder(login.run-04.haas-59.pez.pivotal.io) = cgo
```

**Comment by abbyachau** [comment]
Hi @dmikusa-pivotal could you provide us with more details about this change? What is your use case? What were the issues you were seeing? How could we reproduce the issue? How would this change address those issues? Thanks.

**Comment by dmikusa-pivotal** [comment]
>What is your use case?

In MacOS, there is a feature with DNS resolution where you can add a file to `/etc/resolver/`. The name of the file is the DNS segment and inside the file you specify the DNS resolver to use for that name segment.

Thus if I want to use Google DNS for all google domains, I could add a file `/etc/resolver/google.com` and inside that put `nameserver 8.8.8.8`.  If I then issued a `curl www.google.com`, resolution would be done against 8.8.8.8.

The real use case here is that I want DNS resolution for my CF environment to happen with an internal DNS server, so I put an entry into `/etc/resolver` for my PCF DNS. All works well from my browser or other tools, but the `cf` cli doesn't respect `/etc/resolver/` :(

>What were the issues you were seeing?

See the examples above. DNS resolution fails, but only with he cf cli. It's because it's using Go's built-in resolver, which only takes into consideration `/etc/resolv.conf`.

If you instruct the binary to use the cgo resolver, this delegates to the native resolver libraries on my mac and that takes into account `/etc/resolver/` so everything works :)

Unfortunately, the only way to do that at present is to recompile the cf cli myself. It would be nice if, at least on Mac, the build scripts compiled with cgo by default. Then I could go back to using the officially published binaries.

For what it's worth, I saw docs which seemed to indicate that I could set `GODEBUG=netdns=cgo` and that would instruct the binary to use the cgo resolver, but I did not have any luck making that work with the cf cli. If I could get that working, that would be an acceptable solution too.

>How could we reproduce the issue?

1.) Set up a main DNS server, do that through System Preferences -> Network. Make sure it's unable to resolve the DNS used by your CF environment.
2.) The open a terminal and add a file to `/etc/resolver/`. The name should be the DNS name of your CF environment, ex: `pcf.example.com`. In the file put `nameserver <ip>` and enter the IP of the DNS server which can resolve the DNS for your CF environment.
3.) Run `cf api api.system.pcf.example.com`. DNS resolution will fail because it only takes into account the DNS server you configured in step #1, which cannot resolve the DNS names for your CF environment.

**Comment by willmurphyscode** [comment]
@dmikusa-pivotal We originally switched from permitting cgo to forcing netgo because users who were installing root certs for their own certificate authority would receive untrusted cert errors from macOS 's DNS resolution. See https://www.pivotaltracker.com/n/projects/892938/stories/156851937

We are removing the build time constraint to use netgo, so that macOS CLI will be able to use macOS's DNS resolving behavior, but if this causes the above tracker issue to regress, we will have to choose which issue to fix, or else build two binaries or something.

**Comment by dmikusa-pivotal** [comment]
Ok, let me know when this is available and I can test. I work with labs using self-signed certs too, so if that issue resurfaces, it'll probably hit me too. Thanks!

**Comment by abbyachau** [comment]
@dmikusa-pivotal thanks, the edge binary includes the fix: https://github.com/cloudfoundry/cli/wiki

**Comment by dmikusa-pivotal** [comment]
@abbyachau - I did some tests this morning and it appears to be working now.  I see `go package net: using cgo DNS resolver`. I also fired up Wireshark and I can see the DNS requests going off to the correct DNS servers.

I also do not see any issues with trust validation. I have connected to PWS, which has certs trusted by well-known CAs, one of my labs which has self-signed certs (generated by Ops Manager), and one of my labs that has CAs which are signed by an internal CA (the internal CA is installed in KeyChain). All of them worked as expected.

Hope that helps!
